src/ngraph/runtime/gpu/cublas_emitter.cpp

```diff
@@ -0,0 +1,240 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+
+#include "ngraph/log.hpp"
+#include "ngraph/runtime/gpu/cublas_emitter.hpp"
+#include "ngraph/runtime/gpu/gpu_emitter.hpp"
+#include "ngraph/runtime/gpu/gpu_invoke.hpp"
+#include "ngraph/runtime/gpu/gpu_primitive_emitter.hpp"
+#include "ngraph/runtime/gpu/gpu_runtime_context.hpp"
+#include "ngraph/runtime/gpu/gpu_util.hpp"
+#include "ngraph/runtime/gpu/type_info.hpp"
+#include "ngraph/util.hpp"
+
+using namespace ngraph;
+using namespace std;
+
+runtime::gpu::CUBLASEmitter::CUBLASEmitter(GPUPrimitiveEmitter* emitter, GPURuntimeContext* ctx)
+    : m_primitive_emitter(emitter)
+{
+    m_ctx = ctx;
+}
+
+size_t runtime::gpu::CUBLASEmitter::build_dot(const element::Type& dtype,
+                                              const Shape& arg0_shape,
+                                              const Shape& arg1_shape,
+                                              const Shape& out_shape,
+                                              size_t reduction_axes)
+{
+    std::stringstream ss;
+    ss << "dot_op"
+       << "_dtype_" << dtype.c_type_string() << "_reduction_axes_count_" << reduction_axes;
+    std::string hash = ss.str() + "_i_" + join(arg0_shape, "_") + "_i_" + join(arg1_shape, "_") +
+                       "_o_" + join(out_shape, "_");
+
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    std::unique_ptr<gpu::primitive> dot;
+    if (arg0_shape.empty() || arg1_shape.empty())
+    {
+        auto& second = (arg0_shape.empty() ? arg1_shape : arg0_shape);
+        int count = shape_size(second);
+
+        size_t firstIndex = (arg0_shape.empty() ? 0 : 1);
+        size_t secondIndex = (arg0_shape.empty() ? 1 : 0);
+
+        dot.reset(new gpu::primitive{[=](void** inputs, void** outputs) {
+            CUBLAS_SAFE_CALL(cublasScopy(*m_ctx->cublas_handle,
+                                         count,
+                                         static_cast<const float*>(inputs[secondIndex]),
+                                         1,
+                                         static_cast<float*>(outputs[0]),
+                                         1));
+            CUBLAS_SAFE_CALL(cublasSscal(*m_ctx->cublas_handle,
+                                         count,
+                                         static_cast<const float*>(inputs[firstIndex]),
+                                         static_cast<float*>(outputs[0]),
+                                         1));
+            debug_sync();
+        }});
+
+        return getPrimitiveIndex(dot, hash);
```

Comment:
Can we have a single return at the end, and change this conditional to if-else if?